### PR TITLE
feat: re-implement parseDomain atop psl

### PR DESCRIPTION
**Pull request by FauxFaux**
This PR switches out the `parse-domain` library for `psl` (a popular, supported library without insane dependencies), and a `for` loop. This is as an alternative to #26, which upgrades `parse-domain` to a supported version, at the expense of the option for supporting private subdomains.

This has the advantage of not having to delete any tests!

I believe this loop is a faithful representation of everything parse-domain is doing for the array case.
https://github.com/peerigon/parse-domain/blob/73a8b823b1c6535d7681e80a8667b70941c6568e/src/normalize.js#L17-L19
https://github.com/peerigon/parse-domain/blob/69162fc8430fbffc04f1c3fcf0ce6e9c45e8a8f8/src/parseDomain.js#L14-L20

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -57,18 +61,34 @@ function applyNonce(req: Request, cspString: string): string {
...
+function parseDomain(hostname: string, domainOptions?: ParseOptions): string | null {
+	for (const tld of (domainOptions?.customTlds || [])) {
+		if (hostname.endsWith(`.${tld}`)) {
+			return tld;
+		}
+	}
+
+	let domain = psl.parse(hostname);
+
+	if (undefined !== domain.error) {
```
Comment (frux) [inline src/index.ts]:
```suggestion
	if (domain.error) {
```

**Review by frux — CHANGES_REQUESTED** [review]
It's almost perfect! Here are some minor changes related to codestyle. As you mentioned `parse-domain` allows you to provide any regexp. Lets support this feature. I believe it's easy. In fact it's just one more condition inside `parseDomain`.
```ts
if (customTlds instanceof RegExp) {
	let tld = hostname.match(customTlds);
	if (tld !== null) {
		return tld[0];
	}
}
```

File: src/index.ts [hunk 2]
```diff
@@ -57,18 +61,34 @@ function applyNonce(req: Request, cspString: string): string {
...
 
-        return cspString.replace(new RegExp(TLD, 'g'), domain.tld);
+        return cspString.replace(new RegExp(TLD, 'g'), tld);
     }
 
     return cspString;
 }
 
+function parseDomain(hostname: string, domainOptions?: ParseOptions): string | null {
+	for (const tld of (domainOptions?.customTlds || [])) {
```
Comment (frux) [inline src/index.ts]:
```suggestion
	let customTlds = domainOptions?.customTlds || [];
	for (let tld of customTlds) {
```

**Review by frux — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 2 above)
Comment (FauxFaux) [reply src/index.ts]: This has been lifted into the `else` for the regexp support.

**Comment by FauxFaux** [comment]
It appears `parse-domain` does something odd with leading dots in the regex, so we need to fix up the detected `tld` afterwards, which is a bit sad from a code cleanliness point of view, but doesn't seem particularly risky?

I have added an additional string manipulation to just remove the leading dots, if present. And a test which covers both sides.

File: src/index.ts [hunk 3]
```diff
@@ -57,18 +61,42 @@ function applyNonce(req: Request, cspString: string): string {
...
 }
 
+function parseDomain(hostname: string, domainOptions?: ParseOptions): string | null {
+	let customTlds = domainOptions?.customTlds;
+	if (customTlds instanceof RegExp) {
+		const tld = hostname.match(customTlds);
+		if (tld !== null) {
+			return tld[0].replace(/^\.+/, '');
+		}
+	} else if (customTlds) {
```
Comment (frux) [inline src/index.ts]:
```suggestion
	}

	if (Array.isArray(customTlds)) {
```

**Review by frux — CHANGES_REQUESTED** [review]
Looks great! Here is the last codestyle change from me and I'll publish it.

**Comment by frux** [comment]
I've prepared update of readme #28

File: src/index.ts (hunk 3 above)
Comment (FauxFaux) [reply src/index.ts]: Hah! The code looked like that to start with, but then I changed it to this (to be more similar to what was there before). Fixed.

**Review by frux — APPROVED** [review]

**Comment by frux** [comment]
@FauxFaux I've just published 4.0.0. Thank you for great job!

**Pull request merged**